src/pudl/transform/ferc1.py

```diff
@@ -6509,3 +6540,135 @@ def plants_steam_ferc1(
+    """Create calculation-compnent table from table-level metadata."""
+    clean_xbrl_metadata_json = kwargs["clean_xbrl_metadata_json"]
+    table_dimensions = kwargs["table_dimensions_ferc1"]
+    # compile all of the calc comp tables.
+    calc_metas = []
+    for table_name, tranformer in FERC1_TFR_CLASSES.items():
+        calc_meta = (
+            tranformer(xbrl_metadata_json=clean_xbrl_metadata_json[table_name])
+            .process_xbrl_metadata_calculations()
+            .assign(table_name=table_name)
+        )
+        calc_metas.append(calc_meta)
+    # squish all of the calc comp tables then add in the implicit table dimensions
+    calc_components = pd.concat(calc_metas).pipe(
+        make_implict_dimensions_explict_in_calculaiton_components, table_dimensions
+    )
+    return calc_components
+
+
+def make_implict_dimensions_explict_in_calculaiton_components(
+    calculation_components: pd.DataFrame, table_dimensions: pd.DataFrame
+) -> pd.DataFrame:
+    """Make all of the implied dimensions in the calculation components explict.
+
+    We are assuming that if a dimension was associted with a calculation component, then
+    that is the only dimension that applies whereas if no dimension was associated with
+    a calculation component, that calculation component should be applied across all
+    relevant dimensions.
+
+    Relate all of the calculation components with all found dimensions from
+    :func:`table_dimensions_ferc1`, unless the calculation components already
+    contained information regarding dimensions.
+    """
+    dimensions = [
+        col
+        for col in table_dimensions.columns
+        if col not in ["table_name", "xbrl_factoid"]
+        and col in calculation_components.columns
+    ]
+    logger.info(f"Adding {dimensions=} into calculation component table.")
+    calc_comps_w_dims = calculation_components.copy()
+    # for each dimension, use split/apply/combine. when there are no dims explict in
+    # the calc components, merge in all of the dims.
+    for dim_col in dimensions:
+        null_dim_mask = calc_comps_w_dims[dim_col].isnull()
+        calc_comps_w_implied_dims = pd.merge(
+            calc_comps_w_dims[null_dim_mask].drop(columns=[dim_col]),
+            table_dimensions[["table_name", "xbrl_factoid", dim_col]]
+            .rename(
+                columns={
+                    "xbrl_factoid": "xbrl_factoid_calculation_component",
+                    "table_name": "table_name_calculation_component",
+                }
+            )
+            .drop_duplicates()
+            .dropna(),
+            on=[
+                "table_name_calculation_component",
+                "xbrl_factoid_calculation_component",
+            ],
+            how="left",
+        )
+        calc_comps_w_explicit_dims = calc_comps_w_dims[~null_dim_mask]
+        calc_comps_w_dims = pd.concat(
+            [calc_comps_w_implied_dims, calc_comps_w_explicit_dims]
```

Comment:
Is there a reason to incrementally concatenate with every iteration of the loop rather than doing one concatenation at the end?

Reply:
we are not making new little dfs each iteration, we are split/apply/combining on the same df each iteration. So for each dimension column (`utility_type` or `plant_function` rn) we split the records w/&w/o values in the dimension column, for the records w/o dimension values, we merge in the dimensions found in `table_dimensions` then combine.

Reply:
So the intention is that the new records which are added in each iteration of the loop are going to be included as inputs into subsequent iterations?  If that's true, does it matter what order we loop over the dimensions in? Could we get different outputs if we did `plant_status` before `utility_type` vs. the opposite ordering?

Reply:
this is a good question but I don't believe the order will effect the outpcome. for each dimension column when the values are null, I am dropping the full column & one-to-many merging on a null dimension column.

Reply:
i'll make a unit test for this!